geopandas/algos.c

```diff
@@ -0,0 +1,111 @@
+
+#include <geos_c.h>
+#include "kvec.h"
+#include "algos.h"
+#include <stdio.h>
+#include <time.h>
+#include <stdbool.h>
+
+typedef char (*GEOSPredicate)(GEOSContextHandle_t handle, const GEOSGeometry *left, const GEOSGeometry *right);
+typedef char (*GEOSPreparedPredicate)(GEOSContextHandle_t handle, const GEOSPreparedGeometry *left, const GEOSGeometry *right);
+
+
+typedef struct
+{
+    size_t n, m;
+    GEOSGeometry **a;
+} geom_vector;
+
+
+/* Callback to give to strtree_query
+ * Given the value returned from each intersecting geometry it inserts that
+ * value (typically an index) into the given size_vector */
+void strtree_query_callback(void *item, void *vec)
+{
+    kv_push(size_t, *((size_vector*) vec), (size_t) item);
+}
+
+
+/* Create STRTree spatial index from an array of Geometries */
+GEOSSTRtree *create_index(GEOSContextHandle_t handle, GEOSGeometry **geoms, size_t n)
+{
+    GEOSSTRtree* tree = GEOSSTRtree_create_r(handle, n);
+
+    for (int i = 0; i < n ; i++)
+    {
+        if (geoms[i] != NULL)
+        {
+            GEOSSTRtree_insert_r(handle, tree, geoms[i], (void*) i);
+        }
+    }
+    return tree;
+}
+
+
+/* Spatial join of two arrays of geometries over the predicate
+ * This creates an index for the right side
+ * Finds all polygons in both sides that intersect with each other
+ * Then filters by the spatial predicate like intersects, contains, etc.
+ * This returns an array of indices in each side that match
+ * Organized in a [left_0, right_0, left_1, right_1, ... ] order
+ */
+size_vector sjoin(GEOSContextHandle_t handle,
+                  GEOSPreparedPredicate predicate,
+                  GEOSGeometry **left, size_t nleft,
+                  GEOSGeometry **right, size_t nright)
+{
+    // clock_t begin, begin_1, begin_2;
+    // clock_t end, end_1, end_2;
+    // double time_spent = 0, time_spent_1 = 0, time_spent_2 = 0;
+
+    size_t l, r;                    // indices for left and right sides
+    GEOSPreparedGeometry* prepared; // Temporary prepared geometry for right side
+    size_vector out;                // output array of matching indices
```

Comment:
The reason that a resizable vector is used is because we do not know in advance how many matches there will be, is that correct? If so, can you put somewhere a small comment about it (eg in the explanation above of how the algorithm works)

(starting a bit more to understand the use of the size_vector, at least I think :-))